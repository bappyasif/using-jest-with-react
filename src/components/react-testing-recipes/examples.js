import React, { useEffect, useState } from "react";
import MapComponent from "./map-comp";

export let HelloComp = ({ name }) => {
    if (name) {
        return <h1>Hello, {name}!</h1>;
    } else {
        return <span>Hey, stranger</span>;
    }
}

// Data Fetching
// Instead of calling real APIs in all your tests, you can mock requests with dummy data.
// Mocking data fetching with “fake” data prevents flaky tests due to an unavailable backend, and makes them run faster
export let UserComp = (id) => {
    const [user, setUser] = useState(null);

    async function fetchUserData(id) {
        const response = await fetch("/" + id);
        setUser(await response.json());
    }

    useEffect(() => {
        fetchUserData(id);
    }, [id]);

    if (!user) {
        return "loading...";
    }

    return (
        <details>
            <summary>{user.name}</summary>
            <strong>{user.age}</strong> years old
            <br />
            lives in {user.address}
        </details>
    );
}

// Mocking Modules
// Some modules might not work well inside a testing environment, or may not be as essential to the test itself. 
// Mocking out these modules with dummy replacements can make it easier to write tests for your own code.
// export let MapComp = () => {
//     return (
//         <LoadScript id="script-loader" googleMapsApiKey="YOUR_API_KEY">
//             <GoogleMap id="example-map" center={props.center} />
//         </LoadScript>
//     );
// }

export let ContactComp = (props) => {
    return (
        <div>
            <address>
                Contact {props.name} via{" "}
                <a data-testid="email" href={"mailto:" + props.email}>
                    email
                </a>
                or on their <a data-testid="site" href={props.site}>
                    website
                </a>.
            </address>
            <MapComponent center={props.center} />
        </div>
    );
}

// Events
export let ToggleComp = (props) => {
    const [state, setState] = useState(false);
    return (
        <button
            onClick={() => {
                setState(previousState => !previousState);
                props.onChange(!state);
            }}
            data-testid="toggle"
        >
            {state === true ? "Turn off" : "Turn on"}
        </button>
    );
}

// Timers
// Your code might use timer-based functions like setTimeout to schedule more work in the future. 
// In this example, a multiple choice panel waits for a selection and advances, timing out if a selection isn’t made in 5 seconds:
export let CardComp = (props) => {
    useEffect(() => {
        const timeoutID = setTimeout(() => {
            props.onSelect(null);
            // props.onSelect('done');
        }, 5000);
        console.log('here here!!')
        return () => {
            clearTimeout(timeoutID);
        };
    }, [props.onSelect]);

    return [1, 2, 3, 4].map(choice => (
        <button
            key={choice}
            data-testid={choice}
            onClick={() => props.onSelect(choice)}
        >
            {choice}
        </button>
    ));
}