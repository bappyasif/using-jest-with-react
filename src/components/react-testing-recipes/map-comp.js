import React from 'react'

function MapComponent(props) {
    return (
        <LoadScript id="script-loader" googleMapsApiKey="YOUR_API_KEY">
            <GoogleMap id="example-map" center={props.center} />
        </LoadScript>
    );
}

export default MapComponent