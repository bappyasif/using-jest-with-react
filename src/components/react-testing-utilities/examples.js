import React, { useEffect, useState } from "react"

export let CounterAsFunctionalComp = () => {
    let [counter, setCounter] = useState(0)

    let handleClick = () => setCounter(counter + 1)

    useEffect(() => document.title = `you clicked ${counter} times`, [counter])

    return (
        <div>
            <p>You clicked {counter} times</p>
            <button onClick={handleClick}>
                Click me
            </button>
        </div>
    );
}

export class CounterAsClassComponent extends React.Component {
    constructor(props) {
        super(props)
        this.state = { count: 0 }
        this.handleClick = this.handleClick.bind(this)
    }

    componentDidMount() {
        document.title = `You clicked ${this.state.count} times`
    }

    componentDidUpdate() {
        document.title = `You clicked ${this.state.count} times`
    }

    handleClick = () => this.setState(state => ({ count: state.count + 1 }))

    render() {
        return (
            <div>
                <p>You clicked {this.state.count} times</p>
                <button onClick={this.handleClick}>
                    Click me
                </button>
            </div>
        );
    }
}