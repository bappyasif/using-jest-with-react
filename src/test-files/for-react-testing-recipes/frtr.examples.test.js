// Setup/Teardown
// For each test, we usually want to render our React tree to a DOM element that’s attached to document
// his is important so that it can receive DOM events. When the test ends, we want to “clean up” and unmount the tree from the document.

// const { render } = require("@testing-library/react");
// const { unmountComponentAtNode } = require("react-dom");
// const { act } = require("react-dom/test-utils");
// const { HelloComp } = require("../../components/react-testing-recipes/examples");
import { ContactComp, HelloComp, UserComp } from "../../components/react-testing-recipes/examples"
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { screen } from "@testing-library/dom";
import MapComponent from "../../components/react-testing-recipes/map-comp";

// A common way to do it is to use a pair of beforeEach and afterEach blocks so that they’ll always run and isolate the effects of a test to itself:
let container = null;
beforeEach(() => {
    // setup a dom element as a render target
    container = document.createElement('div')
    document.body.append(container)
})

afterEach(() => {
    // cleaning up on existing
    unmountComponentAtNode(container)
    container.remove()
    container = null;
})
// keep in mind that we want to execute the cleanup even if a test fails. Otherwise, tests can become “leaky”, and one test can change the behavior of another test. That makes them difficult to debug.

// act()
// When writing UI tests, tasks like rendering, user events, or data fetching can be considered as “units” of interaction with a user interface
// react-dom/test-utils provides a helper called act() that makes sure all updates related to these “units” have been processed and applied to the DOM before you make any assertions
// This helps make your tests run closer to what real users would experience when using your application

// Rendering: Commonly, you might want to test whether a component renders correctly for given props:
describe('simple assertions', () => {
    it('renders with or without a name', () => {
        act(() => render(<HelloComp />, container))
        expect(container.textContent).toBe("Hey, stranger");

        act(() => {
            render(<HelloComp name="Jenny" />, container);
        });
        expect(container.textContent).toBe("Hello, Jenny!");

        act(() => {
            render(<HelloComp name="Margaret" />, container);
        });
        expect(container.textContent).toBe("Hello, Margaret!");
    })
})

// fetching data
describe('fetching data', () => {
    it('renders user data', async () => {
        const fakeUser = {
            name: "Joni Baez",
            age: "32",
            address: "123, Charming Avenue"
        };
        jest.spyOn(global, "fetch").mockImplementation(() =>
            Promise.resolve({
                json: () => Promise.resolve(fakeUser)
            })
        );

        // Use the asynchronous version of act to apply resolved promises
        await act(async () => {
            render(<UserComp id="123" />, container);
        });

        expect(container.querySelector("summary").textContent).toBe(fakeUser.name);
        expect(container.querySelector("strong").textContent).toBe(fakeUser.age);
        expect(container.textContent).toContain(fakeUser.address);
        // expect(screen.getByRole())

        // remove the mock to ensure tests are completely isolated
        global.fetch.mockRestore();
    })
})

// Mocking Modules
// If we don’t want to load this component (Map, ContactComp) in our tests, we can mock out the dependency itself to a dummy component, and run our tests
jest.mock("../../components/react-testing-recipes/map-comp", () => {
    return function DummyMap(props) {
        return (
            <div data-testid="map">
                {props.center.lat}:{props.center.long}
            </div>
        );
    };
});
describe('using a mock function', () => {
    it('should render contact information', () => {
        const center = { lat: 0, long: 0 };
        // act(() => render(<ContactComp name="Joni Baez" email="test@example.com" site="http://test.com" center={center} />, container));
        // act(() => {
        //     render(
        //         <ContactComp
        //             name="Joni Baez"
        //             email="test@example.com"
        //             site="http://test.com"
        //             center={center}
        //         />,
        //         container
        //     );
        // });
        // expect(
        //     container.querySelector("[data-testid='email']").getAttribute("href")
        // ).toEqual("mailto:test@example.com");
    })
})