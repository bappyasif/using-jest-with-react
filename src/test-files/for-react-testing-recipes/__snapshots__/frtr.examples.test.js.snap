// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing should render a greeting 1`] = `"<span>Hey, stranger</span>"`;

exports[`snapshot testing should render a greeting 2`] = `"<h1>Hello, Jenny!</h1>"`;

exports[`snapshot testing should render a greeting 3`] = `"<h1>Hello, Margaret!</h1>"`;
